materialize-snowflake/pipe.go

```diff
@@ -0,0 +1,267 @@
+	"fmt"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+)
+
+type PipeClient struct {
+	cfg             *config
+	base            string
+	token           string
+	httpClient      http.Client
+	insertFilesTpl  *template.Template
+	insertReportTpl *template.Template
+	account         string
+}
+
+func publicKeyFingerprint(publicKey *rsa.PublicKey) (string, error) {
+	der, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("marshalling public key: %w", err)
+	}
+	var hash = sha256.Sum256(der)
+	return fmt.Sprintf("SHA256:%s", base64.StdEncoding.EncodeToString(hash[:])), nil
+}
+
+func NewPipeClient(cfg *config, tenant string) (*PipeClient, error) {
+	httpClient := http.Client{}
+
+	var key, err = cfg.Credentials.privateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	var account = strings.ToUpper(cfg.Account)
+	var user = strings.ToUpper(cfg.Credentials.User)
+	var qualifiedUser = fmt.Sprintf("%s.%s", account, user)
+
+	fingerprint, err := publicKeyFingerprint(key.Public().(*rsa.PublicKey))
+	if err != nil {
+		return nil, err
+	}
+
+	var claims = &jwt.RegisteredClaims{
+		IssuedAt: jwt.NewNumericDate(time.Now()),
+		// TODO: automatically refresh the JWT token
+		// JWT tokens for Snowflake can live up to an hour
+		ExpiresAt: jwt.NewNumericDate(time.Now().Add(59 * time.Minute)),
+		Issuer:    fmt.Sprintf("%s.%s", qualifiedUser, fingerprint),
+		Subject:   qualifiedUser,
+	}
+
+	var t = jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	jwtToken, err := t.SignedString(key)
+	if err != nil {
+		return nil, fmt.Errorf("signing key: %w", err)
+	}
+
+	var dsn = cfg.ToURI(tenant)
+	dsnURL, err := url.Parse(fmt.Sprintf("https://%s", dsn))
+	if err != nil {
+		return nil, fmt.Errorf("parsing snowflake dsn: %w", err)
+	}
+
+	var insertFilesTpl = template.Must(template.New("insertFiles").Parse(insertFilesRawTpl))
+	var insertReportTpl = template.Must(template.New("insertReport").Parse(insertReportRawTpl))
+
+	return &PipeClient{
+		cfg:             cfg,
+		base:            dsnURL.Hostname(),
+		httpClient:      httpClient,
+		token:           jwtToken,
+		insertFilesTpl:  insertFilesTpl,
+		insertReportTpl: insertReportTpl,
+	}, nil
+}
+
+type insertFilesURLTemplate struct {
+	Base     string
+	PipeName string
+	// Used to track the request, we generate a random uuid for this
+	RequestId string
+}
+
+type insertFilesRequest struct {
+	Files []FileRequest `json:"files"`
+}
+
+type FileRequest struct {
+	Path string `json:"path"`
+	Size int    `json:"size"`
+}
+
+type InsertFilesResponse struct {
+	Status    string `json:"responseCode"`
+	RequestId string `json:"requestId"`
+}
+
+const insertFilesRawTpl = "https://{{ $.Base }}/v1/data/pipes/{{ $.PipeName }}/insertFiles?requestId={{ $.RequestId }}"
+const contentType = "application/json;charset=UTF-8"
+const userAgent = "Estuary Technologies Flow"
```

Comment:
I don't think it is necessary, I'm mostly providing it for troubleshooting purposes if at some point Snowlake support needs to be involved, so they can find our requests. I don't think adding the tenant name to the user agent would really help with that purpose, since the tenant is implicit in the URL of the request

Reply:
We do get credit for the Snowflake partner program for including the `Application: <TenantName>_EstuaryFlow` parameter to the SQL connection string. If there was a way to set that for these API requests that would be good, since that will be the primary mechanism that credits are consumed with delta updates + snowpipe. No idea if there's actually a way to do that or not, but it would be good to know.